README.rst
RAZIN - Rigorous algebraic zero impeding numeration
=====

For now, I offer
  a) A subroutine to calculate HNF of a matrice whose determinant is small in absolute value;
  b) PLU decomposition and inversion of a matrice modulo 2^64;
  c) Inversion of an integer upper-triangular matrice with small determinant and positive diagonal;
  d) Python binding to some methods of FLINT, called flint_sage;
  e) A subroutine to count an integer matrice determinant.
 
Where are the files?
^^^^^^^^^^^^^^^^^^^^
See *filelist.txt*

My new algorithm to compute integer matrice determinant: ``fmpz_mat_det_hermitian_decomposition()``

Small-det HNF: ``nmod_mat_HNF_nonsquare()`` and ``nmod_mat_HNF()``

What's the target?
^^^^^^^^^^^^^^^^^^
Mid-range target of the project is fast HNF computation with a new algorithm inspired by W.Stein double-determinant. Python wrapper is minimalistic and only contains functions required to reach the goal or to test/benchmark subroutines/algorithms.

The fastest in the open-source world subroutines to count big matrice determinant and small-det matrice Hermite form are a by-product.

FLINT and GMP data structures and subroutines are used.

Bug?
^^^^
If you find out that some subroutine of Razin produces bad result or crashes, your bug-report is welcome. If it includes the following data: sample code including input data, description of what happens, expected result (in case you think output is wrong). For instance, if you think that fmpz_mat_hermite_form() works incorrectly, provide
  a) your code forming input matrice,
  b) output matrice,
  c) the correct output

Glossary
^^^^^^^^

:FLINT:
    C numerical/matrice library with save-every-penny approach to arithmetic. `Here <http://www.flintlib.org/>`_

:krisk0/razin:
    a research log and software package with quickest subroutines for integer linear algebra 

:NTL:
    C++ numerical/matrice library with gluttonous bigint constructor. `Here <http://shoup.net/ntl/>`_

:Captain Flint: 
    fictional Caribbean adventurer (who said pirate?)

:Stepan Razin: 
    16??-1671, kozak, rebel leader (who said gangster?)


Changelog
10 Mar 2014  Fixed bugs in unimodular.{4,5}way.py, which resulted in
              incorrect test results or abnormal termination

             Fredrik Johansson, thanks for help.

             The secret method _clean_cache() is really hard to find.

20140312-r0 flint.pyx, class fmpz_mat:
             wrapped fmpz_mat_solve() as solve_right();
             added __richcmp__()

            benchmark_solve_right.py: benchmarked solve_right() in flint_sage
             and Sage

            Anybody feels like implementing a good interface to Dixon method?
            Question answered, thanks Fredrick

20140313-r1: fmpq_mat constructor workin

20140314-r0: Some fmpq_mat methods wrapped. benchmark_Dixon_square.py
             terminated. Sage .inverse() is a winner for matrix larger than 170

15 Mar 2014  Spent few hours wrapping methods nmod_mat_

             Tests show that nmod_mat_rref does not preserve matrice
              determinant, and nmod_mat_lu* malfunction for non-prime moduli

             So how do I count HNF over ring? Do I need to use NTL HNF?

             I wonder what fmpz_mat_fflu, fmpz_mat_rref,
              fmpz_mat_rref_fraction_free do

17 Mar 2014  Invented ?new? algorithm to compute HNF modulo small number,
              implemented it. C code (cleaned from debug statements) is 318
              lines long. The subroutine turned to be slower than its analog
              in Sage

20140318-r0  Corrected documentation on fmpz_mat_hermite_form() and
              nmod_mat_HNF() interface

20140318-r1  my modular HNF is at least 2.35 times faster than Sage method doing
              the same (ratio depends on dimension). See file
              benchmark_nmod_mat_HNF-sage.cout

20140319-r0  the fastest in galaxy C code to count HNF modulo lattice slightly
              cleaned and now contains 402 lines including 17 assert lines

             nmod_mat_HNF() interface changed: this subroutine returns 0 iff
              determinant confirmed. This is last interface change for this
              subroutine.

             I mean it. Really last last change of the interface.

19 Mar 2014 NTL HNF benchmarked. Guess who wins?

20 Mar 2014 wrapped fmpz_mat_fflu, tried it on small matrice

            Looks like this function correctly counts determinant

22 Mar 2014 re-implemented W.Stein double-determinant algorithm using faster
             subroutines including Dixon linear solver and my modular HNF

            The new subroutine is faster than Sage. For instance
             n=100 bits=256 max time sage/mine=2.91527295113/2.11966109276

24 Mar 2014 Dixon is sometimes faster and sometimes slower, only in add_column().
            In double_det() it is faster always. Profiling takes time...

27 Mar 2014 Equilibrium found, final version of profile_Sage_hnf_square.py is
             running

28 Mar 2014 Profile/Benchmark failed due to bug (exception at line
             'x[i,0] = x[i,0]/d')
            Will record the matrice that causes the bug and report it to Sage
             devs

28 Mar 2014 Sage is playing tricks on me. The exception only occured once, and
             I did not record the matrice. I restared my program with same
             random generator seed, but the exception did not show up.

            Specialized re-implemented hnf_square() runs faster than original,
             30% faster for n=500, bits=512, 13% faster for n=3000, bits=8

03 Apr 2014 riddle `what is this code doing' partially solved
            3 lines
             rhs = a_prime - (w * x)[0]    # row-by-col multiplication
             alpha = rhs / lhs             # integer division
             x=x + alpha*k                 # vector add-mul
            take more than 4 seconds on average when source matrice (on input
             to HNF_square()) is random with n=2000, bits=8

04 Apr 2014 tmod_mat now not only finds LU decomposition (if it exists), but
             also is capable of inverting the two matrix L,U

06 Apr 2014 while random_matrix(ZZ, dim, dim, algorithm='unimodular') takes
             nearly forever for dim=100, my computation of PLU modulo 2**64
             and subsequent inversion of it is so fast that it does not make
             sense to make it faster (gains in speed of HNF computation will
             be unnoticed).

08 Apr 2014 experimenting with different versions of MPIR/GMP, safely, in
             lmonade prefixed environment.

            My idea was to benchmark integer arithmetic. However, flint-2.4.3
             test failed under gmp 6.0.0. Need to further investigate this,
             before doing benchmarks

09 Apr 2014 The error is harmless --- I don't plan to check if smth is
             invertible modulo 1, so hopefully it should not be a problem if
             I choose to switch to linking FLINT to GMP 6.0.0

            Linking flint_sage to MPIR turned to be harder than I thought
             before, due to bugs in gcc/ld behavior. gcc contrary to
             specification links to standard library rather than library in
             custom directory. After walking around the gcc bug my flint_sage
             links to two versions of gmp*.so (one from MPIR and one that is
             used by Sage), and I suspect that code from the 2nd library is used
             in calculation, which is not what I want.

            Tests show that replacing GMP implementation does not make
             determinant() or solve_right() faster on my computer (at least for
             matrice of size 90 with 100-bit entries)

11 Apr 2014 Under GMP 6.0.0a Sage/Flint is a lot faster than under MPIR 2.7.0.
             Probably everybody except me knows that MPIR is not something to
             be used under Linux. Now I know that, too. MPIR, goodbye.

20140414-r0 New subroutine mod_mat_HNF_nonsquare(matrice,number): matrice can be
             non-square; number should be a divisor of determinant of lattice

profile_Sage_hnf_square.py
#!/usr/bin/python2
# -*- coding: utf-8

# This program is part of RAZIN
# Copyright Денис Крыськов 2014
# Licence: GNU General Public License (GPL)

'''
This program is a fork of benchmark_Sage_hnf_square.py

Records time spent by some stages of Stein double-det algorithm as documented
 in To.do
'''

import sage.all
import sys,time,numpy
try:
 import flint_sage as flint
except:
 print 'you forgot to install flint_sage Python wrapper'
 sys.exit(1)

debug_mode=0

write=sys.stdout.write

fmpq_mat,fmpz_mat=flint.fmpq_mat,flint.fmpz_mat
column_to_fmpq_mat=flint.column_to_fmpq_mat

ZZ=sage.all.ZZ
Integer=sage.all.Integer
import sage.matrix.matrix_integer_dense_hnf
det_given_divisor=sage.matrix.matrix_integer_dense_hnf.det_given_divisor
add_row=sage.matrix.matrix_integer_dense_hnf.add_row
add_column=sage.matrix.matrix_integer_dense_hnf.add_column
pivots_of_hnf_matrix=sage.matrix.matrix_integer_dense_hnf.pivots_of_hnf_matrix
matrix=sage.all.matrix
random_matrix=sage.all.random_matrix

t_sage_min=t_sage_max=0
t_mine_min=t_mine_max=0
bits_choice=[8,32,128,256,512]
#           dim  min_bits max_bits
dim_data=[
          ( 100,     8,     512   ),
          ( 250,     8,     512   ),
          ( 500,     8,     256   ),
          (1000,     8,     128   ),
          (2000,     8,       8   ), # with dim=4000 program got killed
          (3000,     8,       8   )] #  ?out of memory?
five=5
if debug_mode:
 dim_data=[(   4,     8,     512   ),
           (   5,     8,     512   ),
           (   6,     8,     512   )]
 five=55

table_to_print=profile_data=None

def profile( e, t ):
 r=time.time()
 global profile_data
 t=r-t
 try:
  s=profile_data[e]
  profile_data[e] = s+t
 except:
  profile_data[e] = t
 return r

def reimplemented_solve_right( A, b ):
 if debug_mode:
  sage_r=A.solve_right(b)
 t0=time.time()
 Af=flint.fmpq_mat( (Integer(1), flint.fmpz_mat(A)) )
 bf=flint.fmpq_mat( (Integer(1), flint.fmpz_mat(b)) )
 mine_r=Af.solve_dixon(bf)
 if debug_mode:
  e=mine_r.export_column().column()
  assert e == sage_r
  print 'solve_right(): check positive'
 profile( 'Dixon solver' , t0 )
 return mine_r

def reimplemented_double_det(A, b, c):
 '''
 nearly identical to double_det(...,proof=True), only uses a faster FLINT 
  method instead of sage solve_right()
 '''
 t0=time.time()
 A = A.transpose()
 b = b.transpose()
 c = c.transpose()
 B = A.augment(b)
 t1=time.time()
 v = reimplemented_solve_right( B, -c )
 profile( 'solve_right-dd' , t1 )
 # ensure that gcd() to canonicalize v and w only runs once
 t1=time.time()
 db = det_given_divisor(B, v.denominator(), proof=True)
 profile( 'det_given_divisor' , t1 )
 n = A.nrows()-1
 vn = v.export_entry(n,0)
 if vn == 0:
  raise ValueError('coin fell on the edge: v[n-1] is zero')
 vn=1/vn
 w=flint.fmpq_mat_scalar_mul_rational( v, -vn )   # w = (-1/vn)*v
 w.entry_mul_Rational( n, vn )  # w[n-1] = w[n-1]/vn
 t1=time.time()
 dc = det_given_divisor(A.augment(c), w.denominator(), proof=True)
 profile( 'det_given_divisor' , t1 )
 if debug_mode:
  assert db==flint.det ( fmpz_mat( B ) )
  assert dc==flint.det ( fmpz_mat( A.augment(c) ) )
  print 'double_det() check positive'
 profile( 'double_det' , t0 )
 return db, dc

def reimplemented_solve_system_with_difficult_last_row(B, a):
 ' return result as fmpq_mat '
 t0=time.time()
 C = sage.all.copy(B)
 D = B.matrix_from_rows(range(C.nrows()-1))
 tK=time.time()
 N = D._rational_kernel_iml()
 profile( 'kernel', tK )
 # original solve_system_with_difficult_last_row() goes into infinite
 #  recursion loop if N.ncols() != 1
 # if this equality ever happens, failed assert is a lot better than infinite
 #  loop
 assert N.ncols() == 1
 k = N.matrix_from_columns([0])
 w = B[-1]
 a_prime = a[-1]
 lhs = w*k
 if lhs[0] == 0:
  # this seldom happens
  # if it happens, original Sage procedure goes into infinite loop
  x=reimplemented_solve_right(B, a)
  if debug_mode:
   assert B*x == a
  return x
 while 1:
  '''
  replace last row of C with random small numbers
  make sure the new matrice is non-singular
  '''
  if debug_mode:
   print 'solve_system_with_difficult_last_row() makin matrice'
  C[C.nrows()-1] = random_matrix( ZZ, 1, C.ncols() ).row(0)
  if quick_nonsigular_test( C ):
   break
 # solve, export as m*1 matrice
 t1=time.time()
 x=reimplemented_solve_right( C, a ).export_column().column()
 profile( 'solve_right-last_row' , t1 )
 rhs = a_prime - w * x
 alpha = rhs[0] / lhs[0]
 x=x + alpha*k
 if debug_mode:
  assert B*x == a
  print 'solve_system_with_difficult_last_row(): salvation correct'
 profile( 'system_with_difficult_last_row' , t0 )
 return x

def reimplemented_add_column( B, H_B, a ):
 ' H_B is Sage matrice or fmpz_mat '
 t0=time.time()
 z = reimplemented_solve_system_with_difficult_last_row(B, a)
 if hasattr(H_B,'nrows'):
  H_B = fmpz_mat( H_B )
 H_Bf=fmpq_mat( (Integer(1), H_B) )
 H_Bf.mul( column_to_fmpq_mat(z.column(0)) )
 r=H_Bf.export_fmpz_mat().export_sage()
 profile( 'add_column' , t0 )
 return r

def reimplement_small_det_HNF(W, g):
 # instead of W._hnf_mod(2*g)
 t0=time.time()
 if g==1:
  r=sage.all.identity_matrix(ZZ, W.nrows())
 else:
  r=flint.fmpz_mat_hermite_form( fmpz_mat( W ), Integer(g) )
 profile( 'small_det_HNF' , t0 )
 return r
 
def reimplemented_hnf_square( A ):
 '''
 A: square non-singular matrice
 
 nearly identical to sage hnf_square()
 '''
 ts=time.time()
 mn = A.nrows()
 B = A.matrix_from_rows(range(mn-2)).matrix_from_columns(range(mn-1))
 c = A.matrix_from_rows([mn-2]).matrix_from_columns (range(mn-1))
 d = A.matrix_from_rows([mn-1]).matrix_from_columns (range(mn-1))
 b = A.matrix_from_columns([mn-1]).matrix_from_rows(range(mn-2))
 # done slicing
 try:
  d1,d2 = reimplemented_double_det( B, c, d )
 except (ValueError, ZeroDivisionError), msg:
  d1 = flint.det ( fmpz_mat_t( B.stack(c) ) )
  d2 = flint.det ( fmpz_mat_t( B.stack(d) ) )
 g,k,l = d1._xgcd (d2, minimal=True)
 W = B.stack (k*c + l*d)
 if g == 0:
  H = W.echelon_form(algorithm='pari')
 else:
  CUTOFF=2**29
  if g>CUTOFF:
   # should never come here
   f = W.gcd()
   g = g / (f**W.nrows())
   if g<=CUTOFF:
     W0 = (W/f).change_ring(ZZ)
     H = reimplement_small_det_HNF(W0, g)
     H *= f
   else:
    #                        ... or here
    raise NotImplementedError("fallback to PARI!")
  else:
   H = reimplement_small_det_HNF(W, g)
 # if H is fmpz_mat, save a penny and let it be
 x = reimplemented_add_column(W, H, b.stack(matrix(1,1,[k*A[mn-2,mn-1] + 
  l*A[mn-1,mn-1]])))
 if debug_mode:
  xsage=add_column(W, H, b.stack(matrix(1,1,[k*A[mn-2,mn-1] + l*A[mn-1,mn-1]])),
   True)
  assert xsage == x
 # if H is fmpz_mat, convert it to Sage
 if not hasattr(H,'nrows'):
  H = H.export_sage()
 Hprime = H.augment(x)
 pivots = range(mn-1)
 if debug_mode:
  assert pivots == pivots_of_hnf_matrix(Hprime)
 t0=time.time()
 Hprime, pivots = add_row(Hprime, A.matrix_from_rows([mn-2]), pivots,
  include_zero_rows=False)
 t1=profile( 'add_row-0', t0 )
 Hprime, pivots = add_row(Hprime, A.matrix_from_rows([mn-1]), pivots,
  include_zero_rows=False)
 profile( 'add_row-1', t1 )
 if debug_mode:
  assert Hprime == Hprime.matrix_from_rows(range(mn))
 profile( 'total', ts )
 return Hprime

def do_benchmark( m ):
 '''
 if result mismatches, abort
 
 if time is new record, update t_****_***
 '''
 global t_sage_min,t_sage_max,t_mine_min,t_mine_max
 t0=time.time()
 sage_r=sage.matrix.matrix_integer_dense_hnf.hnf_square(m,True)
 t1=time.time()
 mine_r=reimplemented_hnf_square(m)
 t2=time.time()
 if mine_r != sage_r:
  print 'result incorrect, m=\n',m
  print 'sage result=\n',sage_r
  print '  my result=\n',mine_r
  sys.exit(1)
 t0=t1-t0
 if t0<t_sage_min:
  t_sage_min=t0
  if t_sage_max < 0:
   t_sage_max=t_sage_min
 else:
  if t0>t_sage_max:
   t_sage_max=t0
 t1=t2-t1
 if t1<t_mine_min:
  t_mine_min=t1
  if t_mine_max < 0:
   t_mine_max=t_mine_min
 else:
  if t1>t_mine_max:
   t_mine_max=t1

def random_data(dim,bits):
 '''
 returns random square non-singular matrice with entries  
 uniformly distributed in the interval −2**bits .. 2**bits
 '''
 while 1:
  a=sage.all.random_matrix( ZZ, dim, x=-1<<bits, y=1<<bits )
  if quick_nonsigular_test( a ):
   return a
  # with non-zero probability matrice a is non-singular, so we do extra
  #  check
  if flint.det(b):
   return a

def quick_nonsigular_test( m ):
 '''
 returns 1 iff determinant of m modulo p0*p1*p1*p3 if non-zero

 where p0...p3 are largest 64-bit primes
 '''
 primes_tail=[ 'C5', 'AD', 'A1', '4D' ]
 for x in primes_tail:
  if flint.det_modulo_prime( m, Integer( (((1<<56)-1) << 8) ^ int(x,16) ) ):
   return 1
 return 0

def benchmark( dim, bits, tries, experiment_no, col_no ):
 '''
 run 2 algorithms multiple times (not more than tries time), store time

 early-abort after 2 or more tries if t_mine_max>60
 '''
 global t_sage_min,t_sage_max,t_mine_min,t_mine_max,profile_data
 t_sage_max=t_mine_max=-1
 t_sage_min=t_mine_min=1e77
 profile_data=dict()
 for i in range( tries ):
  m=random_data( dim, bits )
  do_benchmark( m )
  if t_mine_max > 60 and i:
   tries=i+1
   print 'n=%s bits=%s time=%.1f  benchmarks done=%s, skipping further tries' \
    % (dim,bits,t_mine_max,tries)
   break
 profile_output( dim, bits, tries )
 save_time( dim, bits, experiment_no, col_no )

def sort_in_increasin_order( p ):
 cmp_2nd=lambda x,y: cmp(x[1], y[1])
 l=p.items()
 l.sort( cmp_2nd )
 return l

def detect_profile_format( d ):
 k=0
 for i in d.keys():
  c=len(i)
  if c>k:
   k=c
 return '%'+str(k)+'s %.4f'

def profile_output( dim, bits, tries ):
 fmt=detect_profile_format( profile_data )
 li=sort_in_increasin_order( profile_data )
 he='********* profile for n=%s bits=%s tries=%s *********' % (dim,bits,tries)
 print '\n\n'+he
 for i in li:
  print fmt % (i[0],i[1])
 print ('*' * len(he))+'\n'

def create_table_to_print( d ):
 ' Initialize table to hold benchmark result '
 i=d[0]
 rows=1+2*len(d)
 cols=2
 for j in bits_choice:
  if j >= i[1] and j <= i[2]:
   cols += 1
 t=numpy.resize( numpy.array( [], dtype=object ), (rows,cols) )
 r0=t[0]
 r0[0],r0[1]='','n'
 col=2
 for j in bits_choice:
  if j >= i[1] and j <= i[2]:
   r0[col]='%s bits' % j
   col += 1
 t[1][0]='Sage'
 t[1+len(d)][0]='re-impl'
 for i in range(rows):
  for j in range(cols):
   if t[i][j] == 0:
    t[i][j] = ''
 return t

def save_time( n, b, exp_no, col ):
 ' fill 2 entries in table_to_print '
 save_time_subr( t_sage_min, t_sage_max, 1+exp_no, col )
 save_time_subr( t_mine_min, t_mine_max, 1+len(dim_data)+exp_no, col )

def save_time_subr( mIn, mAx, row, col ):
 global table_to_print
 if mAx<1:
  fmt='%.2f'
 else:
  fmt='%.1f'
 if mIn==mAx:
  e=fmt % mIn
 else:
  e,e1=fmt % mIn,fmt % mAx
  if e != e1:
   e=e+'-'+e1
 table_to_print[ row, col ] = e

sage.all.set_random_seed('20140320')
table_to_print=create_table_to_print( dim_data )
result_row=0
for i in dim_data:
 table_to_print[ 1+              result_row ][ 1 ] = str( i[0] )
 table_to_print[ 1+len(dim_data)+result_row ][ 1 ] = str( i[0] )
 col = 2
 for j in bits_choice:
  if j >= i[1] and j <= i[2]:
   benchmark( i[0], j, five, result_row, col )
   col += 1
   if not debug_mode:
    print 'n=%s bits=%s max time sage/mine=%s/%s' % (i[0],j,t_sage_max,\
     t_mine_max)
    sys.stdout.flush()
 result_row += 1

def pretty_print_result( t, f0 ):
 c=t.shape[1]
 f=numpy.resize( numpy.array( [], dtype=object ), c )
 for i in range(c):
  f[i]=decide_format( t, i, f0 )
 for i in range( t.shape[0] ):
  for j in range( c ):
   s=f[j][0] % t[i][j]
   w=f[j][1]
   if len(s) == w:
    x=w-len( s.lstrip() )
    if x >= 2:
     y=x>>1
     s=s[y:]+( ' ' * y )
     assert len(s)==w
   write(s)
   if j<c-1:
    write(' ')
  write('\n')

def decide_format( t, col, f ):
 req=0
 for i in range(t.shape[0]):
  x=('%'+f) % t[i][col]
  cur=len(x)
  if cur>req:
   req=cur
 return '%'+str(req)+f,req

if not debug_mode:
 pretty_print_result( table_to_print, 's' )
print '\n\nTest passed'
